src/openrct2/drawing/X8DrawingEngine.cpp

```diff
@@ -403,24 +403,34 @@ void X8DrawingEngine::DrawAllDirtyBlocks()
             uint32_t columns = xx - x;
 
             // Check rows
-            uint32_t yy;
-            for (yy = y; yy < dirtyBlockRows; yy++)
+            uint32_t yy = y;
+            if (CheckRows(yy, y, xx, x, dirtyBlockRows, dirtyBlockColumns, columns, dirtyBlocks))
             {
-                uint32_t yyOffset = yy * dirtyBlockColumns;
-                for (xx = x; xx < x + columns; xx++)
-                {
-                    if (dirtyBlocks[yyOffset + xx] == 0)
-                    {
-                        goto endRowCheck;
-                    }
-                }
+                uint32_t rows = yy - y;
+                DrawDirtyBlocks(x, y, columns, rows);
             }
+        }
+    }
+}
 
-        endRowCheck:
-            uint32_t rows = yy - y;
-            DrawDirtyBlocks(x, y, columns, rows);
+bool X8DrawingEngine::CheckRows(
```

Comment:
yy needs to be passed by reference as it is used in the calling function. xx does not need to be passed into this function. `uint32_t dirtyBlockRows, uint32_t dirtyBlockColumns, uint8_t* dirtyBlocks` do not need to be passed into this function as they are available from the class. `x`, `y` and `columns` are not modified in this function and should be `const`. This function does not need to be a `bool`